tools/@aws-cdk/generate-examples/lib/generate.ts

```diff
@@ -0,0 +1,381 @@
+  }
+
+  throw new Error('If this happens, then reflect.typeRefernce must have a new value');
+}
+
+function getBaseUnionType(types: reflect.TypeReference[]): reflect.TypeReference {
+  for (const newType of types) {
+    if (newType.fqn?.endsWith('.IResolvable')) {
+      continue;
+    }
+    return newType;
+  }
+  return types[0];
+}
+
+function addAssumedVariableDeclaration(type: reflect.Type): Code {
+  let newType = type;
+  if (type.isInterfaceType()) {
+    // find corresponding non-interface type if possible
+    newType = findCorrespondingType(type);
+  }
+  const variableName = lowercaseFirstLetter(newType.name);
+  return new Code(variableName, [new Assumption(newType, variableName), new Import(newType)]);
+}
+
+/**
+ * This function tries to guess the corresponding type to an IXxx Interface.
+ * If it does not find that this type exists, it will return the original type.
+ */
+function findCorrespondingType(type: reflect.Type): reflect.Type {
+  const [start, end] = type.fqn.split('.');
+  if (end.length > 2 && end[0] === 'I' && end[1] === end[1].toUpperCase()) {
+    const newFqn = start + '.' + end.substr(1);
+    // eslint-disable-next-line no-console
+    console.log('newFqn ', newFqn);
+    const newType = type.system.tryFindFqn(newFqn);
+    if (newType) { return newType; }
+  }
+  return type;
+}
+
+/**
+ * Helper function to generate an example value for a map.
+ */
+function exampleValueForMap(context: ExampleContext, map: reflect.TypeReference, name: string, level: number): Code {
+  return Code.concatAll(
+    '{\n',
+    new Code(`${tab(level + 1)}${name}Key: `).append(exampleValue(context, map, name, level + 1)).append(',\n'),
+    `${tab(level)}}`,
+  );
+}
+
+/**
+ * Helper function to generate an example value for a struct.
+ */
+function exampleValueForStruct(context: ExampleContext, struct: reflect.InterfaceType, level: number): Code {
+  if (struct.allProperties.length === 0) {
+    return new Code('{ }');
+  }
+  return Code.concatAll(
+    '{\n',
+    ...(struct.allProperties ?? []).map((p) => {
+      return new Code(`${tab(level + 1)}${p.name}: `).append(
+        removeOptional(
+          addOptional(
+            exampleValue(context, p.type, p.name, level + 1),
+            p.optional,
+          ).append(
+            `,${optionalComment(p.optional)}\n`,
+          ),
+        ),
+      );
+    }),
+    `${tab(level)}}`,
+  );
+}
+
+function addOptional(code: Code, optional: boolean ): Code {
+  if (optional && code.code.indexOf('{\n') !== -1) {
+    return code.inject(' // optional', { afterFirstInstanceOf: '{' });
+  }
+  return code;
+}
+
+function removeOptional(code: Code): Code {
+  const i = Math.max(code.code.lastIndexOf('}, // optional'), code.code.lastIndexOf('], // optional'));
+  const j = Math.max(code.code.lastIndexOf('},'), code.code.lastIndexOf('],'));
+  if (i !== -1 && i===j) {
+    return code.remove({ afterLastInstanceOf: ' //' }).append('\n');
+  }
+  return code;
+}
```

Comment:
This is admittedly pretty hacky. It's not what I started out with. There is a function commented below that would be a cleaner solution but I can't figure out if it is possible. The issue is that I want to identify types that are struct types but I end up missing many union types that have structs. We can chat more about it synchronously but wanted to point your attention here.